### PR TITLE
=pro verify BC against all binary compatible versions

### PR DESCRIPTION
**Pull request by 2m**
This is https://github.com/akka/akka/pull/17374 ported to master.

### CODE REVIEW

File: project/plugins.sbt [hunk 1]
```diff
@@ -17,7 +17,8 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-site" % "0.7.1")
 
 addSbtPlugin("com.typesafe.sbt" % "sbt-osgi" % "0.6.0")
 
-addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.1.7")
+addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.1.7-1-g841f792")
+resolvers += Resolver.url("2m-sbt-plugins", url("https://dl.bintray.com/2m/sbt-plugins/"))(Resolver.ivyStylePatterns)
```
Comment (ktoso) [inline project/plugins.sbt]: When can we release it "proper"? Who can?

File: project/AkkaBuild.scala [hunk 2]
```diff
@@ -361,17 +361,27 @@ object AkkaBuild extends Build {
...
+
+      // check against all binary compatible artifacts
+      val artifacts =
+        for (version <- first23.to(last23).map("2.3." + _))
+          yield organization.value %% id % version
+
+      artifacts.toSet
     }
-    else None
+    else Set.empty
```
Comment (ktoso) [inline project/AkkaBuild.scala]: nice

**Comment by ktoso** [comment]
LGTM, though would like to release the plugin "for real" :)

File: project/plugins.sbt (hunk 1 above)
Comment (2m) [inline project/plugins.sbt]: I know @dotta can. :) But we need to merge https://github.com/typesafehub/migration-manager/pull/64 first.

**Comment by 2m** [comment]
PLS BUILD

**Comment by ktoso** [comment]
@2m should we merge this now?

**Comment by ktoso** [comment]
Or we can't release the plugin still - only got repo admin rights?

**Comment by ktoso** [comment]
@2m what's the status here?

**Comment by patriknw** [comment]
@2m are we waiting for anything here or can we merge this?

File: project/plugins.sbt (hunk 1 above)
Comment (patriknw) [inline project/plugins.sbt]: looks like it has been merged, could we release the proper version?

**Comment by 2m** [comment]
I was working on this yesterday. The only thing that we need here is to release `sbt-mima-plugin` which I am on currently.

**Comment by patriknw** [comment]
thanks

**Comment by patriknw** [comment]
any news here? Note that currently we are checking against 2.4.0. If we are to check against at least latest 2.3.x we need a way to define different mima-filters for different versions.

**Comment by 2m** [comment]
Play 2.4.3 uses Akka 2.3.13, and Play 2.3.10 is using Akka 2.3.4. Spark 1.5.1 is using Akka 2.3.11. It looks like we need to go back as much back as 2.3.4 when checking 2.4.x BC.

**Comment by patriknw** [comment]
To be honest I think those that are using Akka 2.3.4 have serious problems anyway.
If we can use the same mima filter for 2.3.4 I don't mind checking, but trying to reproduce the exact filter that for 2.3.4 is a major undertaking (I think).

**Comment by ktoso** [comment]
+1 for checking `2.3.11`, `2.3.14`, `2.4.0`

**Comment by 2m** [comment]
Addded version conditional BC issue filters. Note that on a subsequent (if any) 2.3.x release `bcIssuesBetween23and24` filters will have to be moved to the new version.

File: project/AkkaBuild.scala [hunk 3]
```diff
@@ -404,17 +404,26 @@ object AkkaBuild extends Build {
...
-        case _ : CrossVersion.Full => id + "_" + scalaVersion.value
-        case CrossVersion.Disabled => id
+      val versions = {
+        val akka23Versions = Seq("2.3.11", "2.3.12", "2.3.13", "2.3.14")
+        val akka24Versions = Seq("2.4.0")
+        scalaBinaryVersion.value match {
+          case "2.11" if !Seq(
+            "akka-cluster-sharding",
+            "akka-cluster-tools",
+            "akka-persistence",
```
Comment (patriknw) [inline project/AkkaBuild.scala]: what does this mean? is it excluding those modules from compat check?

Comment (2m) [inline project/AkkaBuild.scala]: It means no `2.3.x` versions for these modules.

Comment (patriknw) [inline project/AkkaBuild.scala]: but akka-persistence-2.4-SNAPSHOT would still be checked against akka-persistence-2.4.0?

Comment (2m) [inline project/AkkaBuild.scala]: Yes, it will be matched by the next case statement.

Comment (ktoso) [inline project/AkkaBuild.scala]: can we make this a bit more named methods? I also was a bit confused what actually will be tested

Comment (patriknw) [inline project/AkkaBuild.scala]: ah, now I see. please extract to something like
```scala
val newArtifactsInAkka24 = Set(
            "akka-cluster-sharding",
            "akka-cluster-tools",
            "akka-persistence",
            "akka-distributed-data-experimental",
            "akka-persistence-query-experimental")
```

File: project/plugins.sbt [hunk 4]
```diff
@@ -17,7 +17,8 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-site" % "0.7.1")
 
 addSbtPlugin("com.typesafe.sbt" % "sbt-osgi" % "0.6.0")
 
-addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.1.7")
+addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.1.7-32-g3bebac4")
+resolvers += Resolver.url("2m-sbt-plugins", url("https://dl.bintray.com/2m/sbt-plugins/"))(Resolver.ivyStylePatterns)
```
Comment (ktoso) [inline project/plugins.sbt]: If we merge in this form let's have a ticket for updating once MiMa gets a release.

Comment (2m) [inline project/plugins.sbt]: Okay. https://github.com/akka/akka/issues/18765

**Comment by patriknw** [comment]
LGTM

**Comment by patriknw** [comment]
@2m looks like the akka-docs jenkins job failed with unresolved dependency: com.typesafe#sbt-mima-plugin;0.1.7-32-g3bebac4: not found

https://jenkins.akka.io:8498/job/akka-docs/680/console

**Pull request merged**